lua/nvim-treesitter/fold.lua

```diff
@@ -1,28 +1,69 @@
+local api = vim.api
+local utils = require'nvim-treesitter.ts_utils'
+local query = require'nvim-treesitter.query'
 local parsers = require'nvim-treesitter.parsers'
 
 local M = {}
 
-function M.get_fold_indic(lnum)
-  if not parsers.has_parser() or not lnum then return '0' end
+local folds_levels = utils.memoize_by_buf_tick(function(bufnr)
+  local levels_tmp = {}
 
-  local function smallest_multiline_containing(node, level)
-    for index = 0,(node:named_child_count() -1) do
-      local child = node:named_child(index)
-      local start, _, stop, _ = child:range()
+  for _, node in ipairs(query.get_capture_matches(bufnr, "@fold", "fold")) do
+    local start, _, stop, _ = node.node:range()
+    stop = stop + 1
 
-      if start ~= stop and start <= (lnum -1) and stop >= (lnum -1) then
-        return smallest_multiline_containing(child, level + 1)
+    -- This can be folded
+    -- Fold only multiline nodes that are not exactly the same as prevsiously met folds
+    if start ~= stop and not (levels_tmp[start] and levels_tmp[stop]) then
+      if levels_tmp[start] and levels_tmp[stop] then
+        print("folding twice :", start, stop)
       end
+
+      levels_tmp[start] = (levels_tmp[start] or 0) + 1
+      levels_tmp[stop] = (levels_tmp[stop] or 0) - 1
+    end
+
+  end
+
+  local levels = {}
+  local current_level = 0
+
+  -- We now have the list of fold opening and closing, fill the gaps and mark where fold start
+  for lnum=0,api.nvim_buf_line_count(bufnr) do
+    local prefix= ''
+    local shift = levels_tmp[lnum] or 0
+
+    if levels_tmp[lnum] and shift >= 0 then
+      prefix = '>'
     end
 
-    return node, level
+    current_level = current_level + (levels_tmp[lnum] or 0)
+    levels[lnum + 1] = prefix .. tostring(current_level)
   end
 
-  local parser = parsers.get_parser()
+  return levels
+end)
 
-  local _, level = smallest_multiline_containing(parser:parse():root(), 0)
+function M.get_fold_indic(lnum)
+  if not parsers.has_parser() or not lnum then return '0' end
+
+  local buf = api.nvim_get_current_buf()
+
+  local levels = folds_levels(buf) or {}
+
+  return levels[lnum] or '0'
+end
+
+function M.attach(bufnr)
+  local win = vim.fn.win_findbuf(bufnr)[1]
+  api.nvim_win_set_option(win, 'foldmethod', 'expr')
+  api.nvim_win_set_option(win, 'foldexpr', 'nvim_treesitter#foldexpr()')
+end
 
-  return tostring(level)
+function M.detach(bufnr)
+  local win = vim.fn.win_findbuf(bufnr)[1]
+  api.nvim_win_set_option(win, 'foldmethod', nil)
+  api.nvim_win_set_option(win, 'foldexpr', nil)
```

Comment:
Detach/attach will probably not work:

 - here it's not guaranteed that the buffer has still a window
 - it is not clear whether detaching from this buffer should deactivate the foldexpr. It might be that the new buffer still supports `foldexpr`

Reply:
Then let's not have fold as a module.
It will keep the thing simple and still allow users to use it.